app/src/main/java/ch/epfl/sdp/peakar/points/POICache.java

```diff
@@ -0,0 +1,155 @@
+package ch.epfl.sdp.peakar.points;
+
+import android.content.Context;
+import android.util.Log;
+
+import com.google.gson.Gson;
+
+import org.osmdroid.util.BoundingBox;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+
+/**
+ * Class used to cache list of POIs in the surrounding
+ * The list of POI is converted to JSON, the corresponding bounding box
+ * is added to the JSON and everything is saved in the cache folder
+ * of the android device
+ */
+public class POICache {
+    //Constants
+    private static final String CACHE_FILE_NAME = "POIPointsCache.txt";
+
+    private static POICache instance;
+    /*List of POIs contained in the cache*/
+    private static ArrayList<POIPoint> cachedPOIPoints;
+    /*Bounding box saved to cache*/
+    private static BoundingBox cachedBoundingBox;
+
+    /**
+     * Private constructor: the class is a singleton
+     */
+    private POICache(){
+        cachedPOIPoints = null;
+        cachedBoundingBox = null;
+    }
+
+    /**
+     * Get singleton
+     * @return POICache instance
+     */
+    public static POICache getInstance(){
+        if(instance == null)
+            instance = new POICache();
+        return instance;
+    }
+
+    /**
+     * Save downloaded POIs and bounding box to cache file
+     */
+    public void savePOIDataToCache(ArrayList<POIPoint> cachedPOIPoints, BoundingBox cachedBoundingBox, Context context){
+        //Create a new object with all needed information to save in JSON
+        POICacheContent poiCacheContent = new POICacheContent(cachedPOIPoints,cachedBoundingBox);
+        Gson gson = new Gson();
+        //Convert to JSON
+        String serializesCache = gson.toJson(poiCacheContent);
+        saveJson(serializesCache,context);
+    }
+
+    /**
+     * Save the serialized JSON to text file
+     * @param serializedObject string in json format
+     * @param context context
+     */
+    private static void saveJson(String serializedObject,Context context){
+        File cacheDir = context.getCacheDir();
+        File outputFile = new File(cacheDir,CACHE_FILE_NAME);
+        try{
+            FileWriter writer = new FileWriter(outputFile);
+            writer.append(serializedObject);
+            writer.flush();
+            writer.close();
+        }
+        catch(IOException e){
+            Log.e("Exception", "File write failed cache POIPoints: " + e.toString());
+        }
+    }
+
+
+    /**
+     * Read text file and retrieve serialized
+     * @return serialized string
+     */
+    private static String readJSON(Context context){
```

Comment:
I'm not sure it is convenient in this case, but GSON offers a "direct" way to read from files.
Check [here](https://stackoverflow.com/questions/29965764/how-to-parse-json-file-with-gson) for more information.

Reply:
Corrected in next commit

Reply:
Update --> The ReadJSON method is back since i have issue to read/write using GSON methods